### PR TITLE
Add a function to join/enclose areas.

### PR DESCRIPTION
**Pull request by gerritholl**
Add a function to calculate an area enclosing one or more other areas.

 - [x] Closes #306
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/master **/*py | flake8 --diff`` 
 - [x] Fully documented

### CODE REVIEW

**Comment by gerritholl** [comment]
The AppVeyor build appears unrelated to this PR.

**Comment by djhoese** [comment]
This reminds me of `gdal_merge.py`, maybe `merge` is another word for this functionality?

**Comment by djhoese** [comment]
Also what are your thoughts on this being a method on the AreaDefinition itself? We could even overload the `__add__` method so you could do `area1 + area2 + area3` and get the enclosing area.

**Comment by gerritholl** [comment]
Merge could be another word.  I chose enclose because the resulting area includes not only the area covered by the input areas, but also the area in-between.  I borrowed this word from the [portion library](https://github.com/AlexandreDecan/portion/blob/master/README.md).

I'm open to add it as a classmethod/constructor on the `AreaDefinition`, but I'm not sure about overloading `+` because I'm not sure the rules usually associated with addition necessarily hold here.  For example, for regular addition, if `a + b + c == a + c`, then  `a + b == a`.  This does not hold for merging or enclosing areas.

**Comment by djhoese** [comment]
Good points. Maybe `|`, but regardless this is "extra" and not something that needs to be done.

**Comment by gerritholl** [comment]
`|` is like joining sets, which also doesn't really hold, because `A | B` in sets should only include what is in either `A` or `B`, but an enclosure may include areas in neither `A` nor `B`.  I would reserve those operators for other operations which might be more appropriate in the future (for example, returning a `StackedAreaDefinition`).

**Comment by djhoese** [comment]
Alright, fine. I suppose the `|` doesn't completely line up with `set` operations. I was more thinking the "opposite" of an intersection (`&`) where the result would be only the overlapping portions of the image.

I wonder if we can or should come up with notation for these types of operations for a pyresample 2.0 kind of interface.

**Comment by mraspaud** [comment]
How about `union` vs `intersection` for clarity?

**Comment by gerritholl** [comment]
@mraspaud The enclosure is neither a union nor an intersection.  It's the geo-2d equivalent of what pyinterval calls the enclosure.  The union `{2, 4} | {6, 8}` does not include 5, but if we have the intervals (2, 4] and (6, 8], then their span (2, 8] _does_ include 5.  This is the 2-dimensional equivalent of such a span/range/enclosure.

Any actual union or intersection function would either need to be very restrictive on its inputs (such as `combine_area_extents_vertical`), or (allow to) return a `StackedAreaDefinition`, or we would need to allow for non-rectangular and even non-contiguous areas.  The latter option would seem quite involved.

**Review by pnuu — APPROVED** [review]
LGTM.

File: pyresample/geometry.py [hunk 1]
```diff
@@ -2443,3 +2438,44 @@ def _dict_representer(dumper, data):
...
+            if not np.isclose(area.resolution, first.resolution).all():
+                raise ValueError("Inconsistent resolution between areas")
+            largest_extent[0] = min(largest_extent[0], area.area_extent[0])
+            largest_extent[1] = min(largest_extent[1], area.area_extent[1])
+            largest_extent[2] = max(largest_extent[2], area.area_extent[2])
+            largest_extent[3] = max(largest_extent[3], area.area_extent[3])
+
+    return create_area_def(
+            area_id=area_id,
+            projection=first.proj_dict,
```
[the remark this answers is not in the document]
Comment (gerritholl) [reply pyresample/geometry.py:2478]: If I change `first.proj_dict` to `first.crs` the test fails with `TypeError: Wrong type for projection: <class 'pyproj.crs.crs.CRS'>. Expected dict or string.` in `_get_proj_data`, which verifies that the projection is either a dict or a string.

[the remark this answers is not in the document]
Comment (djhoese) [reply pyresample/geometry.py:2478]: :thinking: I think in #332 that should work. Once some of my more recent PRs are merged maybe we can test this idea again.

[the remark this answers is not in the document]
Comment (gerritholl) [reply pyresample/geometry.py:2478]: Should I wait for merger of #332?

[the remark this answers is not in the document]
Comment (djhoese) [reply pyresample/geometry.py:2478]: Merged just now. If you have the time to merge master into this PR and test some things that'd be great. Thanks.

[the remark this answers is not in the document]
Comment (gerritholl) [reply pyresample/geometry.py]: Replaced `proj_dict` by `crs`.

[the remark this answers is not in the document]
Comment (gerritholl) [reply pyresample/geometry.py]: Changed name to `enclose_areas`.

**Review by mraspaud — APPROVED** [review]
LGTM, nice feature!

**Pull request merged**